File: leetcode/0242. Valid Anagram/Solution.swift
// https://leetcode.com/problems/valid-anagram/

class Solution { // slow
    func isAnagram(_ s: String, _ t: String) -> Bool {
        return s.sorted() == t.sorted()
    }
}

class Solution {
    func isAnagram(_ s: String, _ t: String) -> Bool {
        var d = [Int](repeating: 0, count: 26)
        let a = Character("a").asciiValue ?? 0
        for c in s {
            let cv = Character(String(c)).asciiValue ?? 0
            d[Int(cv - a)] += 1
        }
        for c in t {
            let cv = Character(String(c)).asciiValue ?? 0
            d[Int(cv - a)] -= 1
        }

        for c in d {
            if c != 0 {
                return false
            }
        }

        return true
    }
}

File: leetcode/0242. Valid Anagram/Solution.py
# https://leetcode.com/problems/valid-anagram/

class Solution:
    # def isAnagram(self, s: str, t: str) -> bool:
    #     return len(s) == len(t) and sorted(s) == sorted(t)

    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        d = [0] * 26
        
        for c in s:
            d[ord(c) - ord("a")] += 1
        
        for c in t:
            d[ord(c) - ord("a")] -= 1

        for c in d:
            if c:
                return False

        return True

s = Solution()
assert s.isAnagram("anagram", "nagaram") == True
assert s.isAnagram("rat", "car") == False
